src/fs/metadata_log.cc

```diff
@@ -0,0 +1,326 @@
+    }));
+}
+
+future<> metadata_log::flush_curr_cluster() {
+    if (_curr_cluster_buff->bytes_left_after_flush_if_done_now(align_after_flush) == 0) {
+        return flush_curr_cluster_and_change_it_to_new_one();
+    }
+
+    schedule_curr_cluster_flush();
+    return _previous_flushes.get_future();
+}
+
+future<> metadata_log::flush_curr_cluster_and_change_it_to_new_one() {
+    auto next_cluster = _cluster_allocator.alloc();
+    if (not next_cluster) {
+        // Here metadata log dies, we cannot even flush current cluster because from there we won't be able to recover
+        // TODO: ^ add protection from it and take it into account during compaction
+        return make_exception_future(no_more_space_exception());
+    }
+
+    auto append_res = _curr_cluster_buff->append(ondisk_next_metadata_cluster {*next_cluster});
+    assert(append_res == metadata_to_disk_buffer::APPENDED);
+    schedule_curr_cluster_flush();
+
+    // Make next cluster the current cluster to allow writing of next metadata entries before flushing finishes
+    _curr_cluster_buff = make_lw_shared<metadata_to_disk_buffer>(_cluster_size, _alignment, cluster_id_to_offset(*next_cluster, _cluster_size));
+
+    return _previous_flushes.get_future();
+}
+
+void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_vec) {
+    // TODO: for compaction: update used inode_data_vec
+    auto file_size = file.size();
+    if (file_size < new_data_vec.data_range.beg) {
+        file.data.emplace(file_size, inode_data_vec {
+            {file_size, new_data_vec.data_range.beg},
+            inode_data_vec::hole_data {}
+        });
+    } else {
+        cut_out_data_range(file, new_data_vec.data_range);
+    }
+
+    file.data.emplace(new_data_vec.data_range.beg, std::move(new_data_vec));
+}
+
+void metadata_log::cut_out_data_range(inode_info::file& file, file_range range) {
+    file.cut_out_data_range(range, [](const inode_data_vec& data_vec) {
+        (void)data_vec; // TODO: for compaction: update used inode_data_vec
+    });
+}
+
+std::variant<inode_t, metadata_log::path_lookup_error> metadata_log::path_lookup(const sstring& path) const {
+    if (path.empty() or path[0] != '/') {
+        return path_lookup_error::NOT_ABSOLUTE;
+    }
+
+    std::vector<inode_t> components_stack = {_root_dir};
+    size_t beg = 0;
+    while (beg < path.size()) {
+        range component_range = {beg, path.find('/', beg)};
+        bool check_if_dir = false;
+        if (component_range.end == path.npos) {
+            component_range.end = path.size();
+            beg = path.size();
+        } else {
+            check_if_dir = true;
+            beg = component_range.end + 1; // Jump over '/'
+        }
+
+        // TODO: I don't like that we make a copy here -- it is totally redundant and inhibits adding noexcept
+        sstring component = path.substr(component_range.beg, component_range.size());
```

Comment:
1. Please switch to std::string, as mentioned above
2. How about std::string_view then? Using it will postpone the copy until it's really needed (e.g. the substring needs to be put somewhere)